design/ceph-openshift-fencing-mitigation.md

```diff
@@ -0,0 +1,52 @@
+
+## What is fencing in OCP 4.x?
+
+Openshift uses `Machines` and `MachineSets` from the [cluster-api](https://github.com/kubernetes-sigs/cluster-api) to dynamically provisions nodes. Fencing is a remediation method that reboots/deletes `Machine` CRDs to solve problems with automatically provisioned nodes.
+
+Once [MachineHealthCheck controller](https://github.com/openshift/machine-api-operator#machine-healthcheck-controller) detects that a node is `NotReady` (or some other configured condition), it will remove the associated `Machine` which will cause the node to be deleted. The `MachineSet` controller will then replace the `Machine` via the machine-api. The exception is on baremetal platforms where fencing will reboot the underlying `BareMetalHost` object instead of deleting the `Machine`.
+
+
+## Why can't we use `PodDisruptionBudget`?
+
+Fencing does not use the eviction api. It is for `Machine`s and not `Pod`s.
+
+## Will we need to do large storage rebalancing after fencing?
+
+Hopefully not. On cloud platforms, the OSDs can be rescheduled on new nodes along with their backing PVs, and on baremetal where the local PVs are tied to a node, fencing will simply reboot the node instead of destroying it.
+
+
+# Problem Statement
+We need to ensure that only one node can be fenced at a time and that Ceph is fully recovered (has PGs clean) before any fencing is initiated. The available pattern for limiting fencing is the MachineDisruptionBudget which allows us to specify maxUnavailable. However, this won’t be sufficient to ensure that Ceph has recovered before fencing is initiated.
+
+Therefore, we will control how many nodes match the MDB by dynamically adding and removing labels as well as dynamically updating the MDB. By manipulating the MDB into a state where desiredHealthy > currentHealthy, we can disable fencing on the nodes the MDB points to.
+
+# Design:
+
+## machine-controller
+Each `Machine` with OSDs from a `CephCluster` will be labeled with the key `fencegroup.rook.io/<cluster-name> by this controller.
+
+This controller will maintain a label on the `Machine` as long as there are osds scheduled there that are not down. Therefore fencing will only be allowed on nodes with no OSDs unless the MDB has maxUnavailable == 1.
+
+## machinedisruptionbudget-controller
+Whenever a `CephCluster` is created, a corresponding `MachineDisruptionBudget` will be created that is dynamically controlled by this controller. It will point at `Machine`s labeled `fencegroup.rook.io/<cluster-name>`. It has no effect on `Machine`s without the label.
+
+If all PGs are clean, maxUnavailable will be set to 1, otherwise there is a problem and fencing will be blocked by setting maxUnavailable=0.
+
+## Assumptions:
+- We assume that the controllers will be able to reconcile multiple times in < 5 minutes as we know that fencing will happen only after a configurable timeout. The default timeout is 5 minutes.
```

Comment:
What triggers the reconciliation? Is this tied to the ceph status query, which is every 60s?

Reply:
Adding what triggers the reconciliation per-reconciler.